From 5977b7b63e0e2a5472d21452612f2e98294b1891 Mon Sep 17 00:00:00 2001
From: minaripenguin <minaripenguin@users.noreply.github.com>
Date: Fri, 1 Dec 2023 17:16:44 +0800
Subject: [PATCH 40/48] Keystore: Spoof locked bootloader on local attestations

based from: https://github.com/chiteroman/BootloaderSpoofer

Squashed:

    From: minaripenguin <minaripenguin@users.noreply.github.com>
    Date: Tue, 5 Dec 2023 19:56:07 +0800
    Subject: fixup! Keystore: Spoof locked bootloader on local attestations

    * as reported by sir shamik without this fixup, bootloader is detected as unlocked on keystore attestation app.
    * thanks to ghostrider-reborn for the references and clarification about the injection

    reference: https://github.com/RisingTechOSS/android_frameworks_base/commit/d57f22f7d99a29ce860d7113d78eb8122a8b7816#r134217184

    test: m confirmed that key attestation 1.5.0 detects device bootloader state as locked

    Signed-off-by: minaripenguin <minaripenguin@users.noreply.github.com>

Co-authored-by: chiteroman <98092901+chiteroman@users.noreply.github.com>
Signed-off-by: minaripenguin <minaripenguin@users.noreply.github.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../keystore2/AndroidKeyStoreSpi.java         | 40 +++++++++++++++----
 1 file changed, 33 insertions(+), 7 deletions(-)

diff --git a/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java b/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
index 2e16c0c51db4..6818a6dc250a 100644
--- a/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
+++ b/keystore/java/android/security/keystore2/AndroidKeyStoreSpi.java
@@ -87,6 +87,7 @@ import java.util.Date;
 import java.util.Enumeration;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Locale;
 import java.util.NoSuchElementException;
 
 import javax.crypto.SecretKey;
@@ -179,6 +180,20 @@ public class AndroidKeyStoreSpi extends KeyStoreSpi {
         }
     }
 
+    private static int indexOf(byte[] array) {
+        final byte[] PATTERN = {48, 74, 4, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 10, 1, 2};
+        outer:
+        for (int i = 0; i < array.length - PATTERN.length + 1; i++) {
+            for (int j = 0; j < PATTERN.length; j++) {
+                if (array[i + j] != PATTERN[j]) {
+                    continue outer;
+                }
+            }
+            return i;
+        }
+        return -1;
+    }
+
     @Override
     public Certificate[] engineGetCertificateChain(String alias) {
         if (SystemProperties.getBoolean("persist.sys.pihooks.enable", true)) {
@@ -196,15 +211,28 @@ public class AndroidKeyStoreSpi extends KeyStoreSpi {
             return null;
         }
 
-        final Certificate[] caList;
+        X509Certificate modLeaf = leaf;
+        try {
+            byte[] bytes = leaf.getEncoded();
+            if (bytes != null && bytes.length > 0) {
+                int index = indexOf(bytes);
+                if (index != -1) {
+                    bytes[index + 38] = 1;
+                    bytes[index + 41] = 0;
+                    CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+                    X509Certificate modCert = (X509Certificate) certFactory.generateCertificate(new ByteArrayInputStream(bytes));
+                    modLeaf = modCert;
+                }
+            }
+        } catch (CertificateException e) {
+            return null;
+        }
 
         final byte[] caBytes = response.metadata.certificateChain;
-
+        final Certificate[] caList;
         if (caBytes != null) {
             final Collection<X509Certificate> caChain = toCertificates(caBytes);
-
             caList = new Certificate[caChain.size() + 1];
-
             final Iterator<X509Certificate> it = caChain.iterator();
             int i = 1;
             while (it.hasNext()) {
@@ -213,9 +241,7 @@ public class AndroidKeyStoreSpi extends KeyStoreSpi {
         } else {
             caList = new Certificate[1];
         }
-
-        caList[0] = leaf;
-
+        caList[0] = modLeaf;
         return caList;
     }
 
-- 
2.25.1

